README.md
# go seof: Simple Encrypted os.File

Encrypted drop-in replacement of golang' [`os.File`](https://golang.org/pkg/os/#File), the file stored in disk will be
encrypted, and the resulting type can be used anywhere an [`os.File`](https://golang.org/pkg/os/#File) could be used.
i.e. it can be both sequentially and randomly read and write, at any file position for any amount of bytes, can be
truncate, seek, stats, etc. i.e. [`Read`](https://golang.org/pkg/os/#File.Read),
[`ReadAt`](https://golang.org/pkg/os/#File.ReadAt),
[`WriteAt`](https://golang.org/pkg/os/#File.WriteAt),
[`Seek`](https://golang.org/pkg/os/#File.Seek),
[`Truncate`](https://golang.org/pkg/os/#File.Truncate), etc.

It derives a file-wide key using [scrypt](http://www.tarsnap.com/scrypt.html) with a provided string password, the file
is sliced into blocks of n bytes (decided at creation time.). Each block is encrypted and sealed using three AES256/GCM
envelops, one inside the other, with three different keys and nonces achieving
both [confidentiality and authenticity](https://en.wikipedia.org/wiki/Authenticated_encryption). File wide integrity is
warrantied by signing blocks and avoiding empty sparse blocks.

Current Version: [v1.0.0](https://github.com/kuking/seof/tree/v1.0.0), changelog [here](CHANGELOG.md).

Example
-------

Snippet taken from [base_test.go](base_test.go). Check the test files for more examples, i.e. Seek, Truncate, Stats,
etc.

```
    password := "this is a very long password nobody should know about"
    BEBlockSize := 1024
    data := crypto.RandBytes(BEBlockSize*10)

    // create, write, close.
    f, err := seof.CreateExt("encrypted.seof", password, BEBlockSize, 1)
    assertNoErr(err, t)

    n, err := f.Write(data)
    assertNoErr(err, t)
    if n != len(data) {
        t.Fatal("did not write the whole buffer")
    }
    err = f.Close()
    assertNoErr(err, t)

    // open, read, close.
    f, err = seof.OpenExt("encrypted.seof", password, 1)
    assertNoErr(err, t)
    readBuf := make([]byte, BEBlockSize*15) // bigger, purposely
    n, err = f.Read(readBuf)
    if n != len(data) {
        t.Fatal("It did not read fully")
    }
    if !bytes.Equal(data, readBuf[0:n]) {
        t.Fatal("read error, does not equals to initial write")
    }
    err = f.Close()
    assertNoErr(err, t)

```

CLI
---

Usage, it can encrypt/decrypt/inspect files' metadata from CLI:

```$ ./seof                                                                                                                                                                                              ed@luxuriance
Usage of ./seof: seof file utility

  -e	encrypt (default: to decrypt)
  -h	Show usage
  -i	show seof encrypted file metadata
  -p string
    	password file
  -s uint
    	block size (default: 1024)
  -scrypt string
    	Encrypting Scrypt parameters: min, default, better, max (default "default")

NOTES:
  - Password must be provided in a file. Command line is not secure in a multi-user host.
  - When encrypting, contents have to be provided via stdin pipe, decrypted output will be via stdout.
  - Scrypt parameters target times in modern CPUs (2021): min>20ms, default>600ms, better>5s, max>9s

Examples:
  $ cat file | seof -e -p @password_file file.seof
  $ seof -p @password_file file.seof > file
  $ seof -i -p @password_file file.seof
```

Inspecting metadata for an encrypted file:

```
$ /seof -p password -i file.seof                                                                                                                                                                     ed@luxuriance
           File Name: file.seof
   Modification Time: 2021-01-03 13:53:55.698769333 +0000 GMT
           File Mode: -rw-r--r--
        Content Size: 247086468 bytes
   File Size On Disk: 268321756 bytes
 Encryption Overhead: 8.59%
  Content Block Size: 1024 bytes
Encrypted Block Size: 1112 bytes
 Total Blocks Writen: 241298 (= unique nonces)
       SCrypt Preset: Maximum (>9s)
   SCrypt Parameters: N=524288, R=64, P=1, keyLength=96, salt=
     e036b1c8443913266fa514404dc56fa2603e5215136dfe7b83cb2149eb924dc1
     40cc023e94fcde57b4ca095e81b3ab94331a9defbb03187b4a1761ee37179402
     f206d9f768034a9cb7d42e9355f55876c4ffb8710da32d56c6b384101a3d13f4
  ```

Performance
-----------
There is no performance overhead beyond the encryption primitives. Internally, `seof` holds multiple unencrypted blocks
in memory, unbuffered reading and writing should not incur in any extra encryption work, and the typical sequential
reads and writes should be performant independently of the access pattern.

Finally, encryption occurs in blocks, so changing just one byte would require encrypting and storing a whole block (i.e.
10kb). You want to tune the quantity of in-memory blocks when opening the file; and the block size when creating it.

### CLI sequential encryption/decryption performance

(MacBook Pro (13-inch, 2018, Four Thunderbolt 3 Ports), 2.7 GHz Quad-Core Intel Core i7)

```
$ cat ~/Downloads/debian-10.5.0-amd64-netinst.iso | pv | ./seof -p password -e debian-10.5.0-amd64-netinst.iso.seof 
 349MiB 0:00:06 [50.8MiB/s] [                            <=>                                                    ]

$ ./seof -p password debian-10.5.0-amd64-netinst.iso.seof | pv > debian-10.5.0-amd64-netinst.iso                                                                                                      ed@luxuriance
 349MiB 0:00:02 [ 132MiB/s] [            <=>                                                                    ]
```

File Structure
--------------

- Header: (128 bytes, 120 used)
    - uint64 Magic
    - [96]byte Script salt
    - uint32 Scrypt parameters: N, R, P.
    - uint32 Disk block size
    - [8]byte zeros (verified on open)
- A block:
    - [36]byte: nonce
    - uint32: cipherText length
    - [disk-block-size]byte: CGM stream
        - the additional data for the AEAD is an uint64 holding the block number (verified)
- Special block 0:
    - uint64: File size
    - uint32: Disk block size (must eq to the header)
    - uint32: un-encrypted block size
    - uint64: written blocks (as in number of unique nonces generated)
    - []byte: Further metadata expansion

Testing
-------
Code is extensively tested and there is a soak test suite that tests multiple access patterns (i.e. misaligned reads,
writes, multi-blocks ops, sub-block ops, concurrency, etc).

```
$ ./soaktest                                                                                                                                                                                          ed@luxuriance
soaktest: seof soak test, creates a native file and a seof encrypted file.
  applies many different IO operations equally to both files and verifies both behave similar. You want a fast disk (NVMe).

1. Creating 2 x 256MB files: native.soak, seof.soak
2. Writing 256MB of [0x00, 0x01, 0x02, ... 0xff] in: native.soak, seof.soak
.................................................. done
3.1. Fully comparing files, using read_chunk_size=1
.................................................. done
3.2. Fully comparing files, using read_chunk_size=2
.................................................. done
3.3. Fully comparing files, using read_chunk_size=3
.................................................. done
3.4. Fully comparing files, using read_chunk_size=4
.................................................. done

[...]

.................................................. done
3.16. Fully comparing files, using read_chunk_size=16
.................................................. done
3.17. Fully comparing files, using read_chunk_size=256
.................................................. done
3.18. Fully comparing files, using read_chunk_size=512
.................................................. done
3.19. Fully comparing files, using read_chunk_size=924
.................................................. done
3.20. Fully comparing files, using read_chunk_size=1023
.................................................. done
3.21. Fully comparing files, using read_chunk_size=1024
.................................................. done
3.22. Fully comparing files, using read_chunk_size=1025
.................................................. done
3.23. Fully comparing files, using read_chunk_size=1124
.................................................. done
3.24. Fully comparing files, using read_chunk_size=2048
.................................................. done
3.25. Fully comparing files, using read_chunk_size=3072
.................................................. done
3.26. Fully comparing files, using read_chunk_size=4096
.................................................. done
3.27. Fully comparing files, using read_chunk_size=4095
.................................................. done
3.28. Fully comparing files, using read_chunk_size=4097
.................................................. done
4.1.1. Rewriting wholy using chunk_size=1
.................................................. done
4.1.2. Verifying (fast, using chunk_size=1024)
.................................................. done
4.2.1. Rewriting wholy using chunk_size=2
.................................................. done
4.2.2. Verifying (fast, using chunk_size=1024)
.................................................. done
4.3.1. Rewriting wholy using chunk_size=3
.................................................. done

[...]

4.22.1. Rewriting wholy using chunk_size=1025
.................................................. done
4.22.2. Verifying (fast, using chunk_size=1024)
.................................................. done
4.23.1. Rewriting wholy using chunk_size=1124
.................................................. done
4.23.2. Verifying (fast, using chunk_size=1024)
.................................................. done
4.24.1. Rewriting wholy using chunk_size=2048
.................................................. done
4.24.2. Verifying (fast, using chunk_size=1024)
.................................................. done
4.25.1. Rewriting wholy using chunk_size=3072
.................................................. done
4.25.2. Verifying (fast, using chunk_size=1024)
.................................................. done
4.26.1. Rewriting wholy using chunk_size=4096
.................................................. done
4.26.2. Verifying (fast, using chunk_size=1024)
.................................................. done
4.27.1. Rewriting wholy using chunk_size=4095
.................................................. done
4.27.2. Verifying (fast, using chunk_size=1024)
.................................................. done
4.28.1. Rewriting wholy using chunk_size=4097
.................................................. done
4.28.2. Verifying (fast, using chunk_size=1024)
.................................................. done
5.1. Writing 262144 random chunks of miscelaneous sizes of up to 2048 bytes
................................................... done
5.2. Verifying (fast, using chunk_size=1024)
.................................................. done
6.1. Reading 262144 random chunks of miscelaneous sizes of up to 2048 bytes
................................................... done
7.1 Synchronisation: reading native, writing encrypted 1048576 chunks of up to 2048 bytes within 64 concurrent threads
.................................................. done
7.2. Verifying (fast, using chunk_size=1024)
.................................................. done
7.3. Synchronisation: reading encryptede 1048576 chunks of up to 2048 bytes within 64 concurrent threads
.................................................. done

SUCCESS!

```

Syncronisation
--------------
Concurrency safety is achieved with a global lock, do not expect optimal concurrent performance. It is safe to do
operations on the same seof File object from multiple concurrent goroutines.

Attack vectors
--------------

- Each time a block is written, a new **random** nonce is generated. Internally, the implementation uses buffers and
  will write to disk only when the buffer needs to be flushed (file closed, sync or cache eviction.). It is a
  requirement for GCM to never reuse a nonce, or the key can be compromised. We have calculated the odds of duplicating
  a nonce (consisting of 12 bytes of randomness), see the details in the
  ticket [here](https://github.com/kuking/seof/issues/1) and
  the [calculator](https://github.com/kuking/seof/blob/master/cli/birthday/main.go). Long story short, with triple-AES
  is practically impossible, with single AES, the chance is 1 in a billion after writing 37TiB into one single file. If
  you are worried about those odds, create multiple smaller files, the password can be reused as the scrypt will be
  initialised with different salts in each file. To put this number in perspective, the average write-life expectancy
  for a modern SSD disk is 500TiB. Finally, special block 0 holds a counter with the number of unique nonces ever
  generated. This value can be inspected using the `seof -i` CLI command or via the `Stats` function.

- The weakest encryption-link is the password string used for generating the 768 bits (96 bytes) of key. A string in
  latin characters should have to be approx. 150 characters in order to hold 768 bits of entropy. You have to keep that
  in mind.

- Blocks within the same file can not be shuffled or moved to another block (or even another file) as the AEAD seals
  hold the block number in the signed plaintext. This is verified.

- Replacing a ciphertext block with a previous copy of the same block will not be detected. Like replacing the whole
  encrypted file with a previous version of itself. For the user this will experienced as if the file as lost written
  data (as the previously stored data would come back). An attacker has to have read/write access to the filesystem, but
  will not be able to generate new arbitrary plaintext. It is possible to prevent this attack, at both performance cost
  and higher risk of corrupting the file on failure scenarios (i.e. block is flushed, but reference to block
  high-water-mark is lost).

- Most filesystems can handle [sparse files](https://en.wikipedia.org/wiki/Sparse_file). seof supports sparse files, but
  read of never written/zeroed blocks is disabled by default to avoid a possible attack (see: XXX flag). User can create
  a new file and [`Seek`](https://golang.org/pkg/os/#File.Seek) to any part of it, write a byte, and later read it.
  Reading outside the block boundaries of the unique written byte will fail unless explicitly enabled. This is not a
  very typical user case.

  __Long explanation__: in order to keep track of blocks holding data, seof should keep a block-written-bitmap. So when
  a block is read from the disk and comes completely empty (zeroed, no AEAD seal present), but the block-written-bitmap
  accuses it was written previously, it is fair to assume the data has been lost, therefore deemed inconsistent, an IO
  error should be raised (it could have been zeroed by a malicious actor, too.). Without this block-written-bitmap, a
  zeroed block by a malicious actor and an honest empty blob in a sparse file are indistinguishable, potentially
  allowing a "selective block zero-ing attack." and failing the integrity assurances.

  If you really need this assurance, let me know, the block-written-bitmap can be done.

USAGE
-----

- Storing passwords and secrets using auto-generated system+app+user derived key
- Encrypting distributable assets and you need random access reads. (i.e. reading a ZIP File)
- Enhancing encryption in traditional file formats (i.e.
  golang' [zip reader](https://golang.org/pkg/archive/zip/#NewReader))
- Secure long-term storing of files (some people might want to use GPG as it is "proven" to work)
- Keeping usage data away from user's eyes
- Random access on *very* big files, seof supports 64bits files. i.e. efficient and fast random access of >4gb files.

- Any of the above and you really want to make it future proof, i.e. scenario where AES is degraded.

TODO
----

- Flag to allow reading empty holes in sparse files as no errors
- Crypto analysis


CHANGELOG.md
# Changelog

## v1.0.1
2023-06-30

### Notes
- Updated x/crypto to latest version
- Updated Hashicorp LRU library to latest version that does not require generics

## v1.0.0
2021-01-05

### Notes

- Feature complete
- Production ready (as per license, you are using it at your own risk.)
- Multithreading safe
- Quality assurance:
    - Unit tests covers ~ 87% of code
    - Extensive soak test, including multithreading tests and multiple (mis)aligned access patterns
    - No linting issues


api_test.go
package seof

import (
	"github.com/kuking/seof/crypto"
	"os"
	"testing"
)

const (
	password    = "4f760f9e67c61ff63044fa97e0b00fdc96c744a3"
	BEBlockSize = 1024
)

func TestNoUsableMethods(t *testing.T) {
	if _, err := Create("any"); err == nil {
		t.Fatal("Create should not work, use CreateExt")
	}
	if _, err := Open("any"); err == nil {
		t.Fatal("Open should not work, use OpenExt")
	}
	if _, err := OpenFile("any", 0, os.ModeAppend); err == nil {
		t.Fatal("OpenFile should not work, use OpenExt")
	}
}

func TestCreateExt_InvalidArguments(t *testing.T) {
	for _, password := range []string{"", "1234567890", "1234567890123456789"} {
		if _, err := CreateExt("file", password, crypto.MinSCryptParameters, BEBlockSize, 1); err == nil {
			t.Fatal("password should be at least 20 characters long")
		}
	}
	for _, blockSize := range []int{-123, -1, 0, 10, 1023, 128*1024 + 1, 256 * 1024} {
		if _, err := CreateExt("file", password, crypto.MinSCryptParameters, blockSize, 1); err == nil {
			t.Fatal("block size should be: 1kb<=block_size<128kb")
		}
	}
	for _, memBuffers := range []int{-123, -1, 0, 129, 65535} {
		if _, err := CreateExt("file", password, crypto.MinSCryptParameters, BEBlockSize, memBuffers); err == nil {
			t.Fatal("memory buffers be: 1<=buffers<128")
		}
	}
	_, err := CreateExt("", password, crypto.MinSCryptParameters, BEBlockSize, 1)
	if err == nil {
		t.Fatal("invalid filename should fail")
	}
}

func TestOpenExt_InvalidArguments(t *testing.T) {
	for _, buffers := range []int{-123, -1, 0, 1025, 128 * 1024} {
		_, err := OpenExt("file", password, buffers)
		if err == nil {
			t.Fatal("memory buffers should be checked for bounds")
		}
	}
	_, err := OpenExt("", password, 1)
	if err == nil {
		t.Fatal("invalid filename should fail")
	}
}


crypto_test.go
package seof

import (
	"github.com/kuking/seof/crypto"
	"testing"
)

func TestSealOpen(t *testing.T) {
	f := File{}
	h := givenValidHeader()
	_ = f.initialiseCiphers(password, &h)

	plainText := "This is a secret"
	cipherText, nonce := f.seal([]byte(plainText), 1234)
	recoveredText, err := f.unseal(cipherText, 1234, nonce)
	if err != nil {
		t.Fatal(err)
	}
	if string(recoveredText) != plainText {
		t.Fatal("recovered plaintext not equal")
	}
}

func TestSealOpen_InvalidBlockNo(t *testing.T) {
	f := File{}
	h := givenValidHeader()
	_ = f.initialiseCiphers(password, &h)
	plainText := "This is a secret"
	cipherText, nonce := f.seal([]byte(plainText), 1234)
	_, err := f.unseal(cipherText, 5432, nonce)
	if err == nil {
		t.Fatal(err)
	}
}

func TestSealOpen_Sizes(t *testing.T) {
	f := File{}
	h := givenValidHeader()
	_ = f.initialiseCiphers(password, &h)
	plainText := "This is a secret"
	cipherText, nonce := f.seal([]byte(plainText), 1234)

	if len(nonce) != 36 || len(nonce) != nonceSize {
		t.Fatal("nonce has to be 12*3 bytes")
	}
	if float32(len(plainText))*1.5 > float32(len(cipherText)) {
		t.Fatal("cipherText seems too short")
	}
}

// of course we don't intend to test the crypto primitives here, we want to assert without any doubt we did not "f.up"
// the integration with the crypto primitives, now or in the future.
func TestSealOpen_AnyByteChangeShouldFail(t *testing.T) {
	f := File{}
	h := givenValidHeader()
	_ = f.initialiseCiphers(password, &h)
	plainText := "This is a secret"
	cipherText, nonce := f.seal([]byte(plainText), 1234)
	// cipher-text
	for i := 0; i < len(cipherText); i++ {
		orig := cipherText[i]
		cipherText[i] = crypto.RandBytes(1)[0]
		if cipherText[i] == orig {
			cipherText[i]++
		}
		_, err := f.unseal(cipherText, 1234, nonce)
		if err == nil {
			t.Fatal("this should have failed after changing one byte in the cipherText")
		}
		cipherText[i] = orig
	}
	// nonce
	for i := 0; i < len(nonce); i++ {
		orig := nonce[i]
		nonce[i] = orig ^ crypto.RandBytes(1)[0]
		if nonce[i] == orig {
			nonce[i]++
		}
		_, err := f.unseal(cipherText, 1234, nonce)
		if err == nil {
			t.Fatal("this should have failed after changing one byte in the cipherText")
		}
		nonce[i] = orig
	}
}


structs.go
package seof

import (
	"bytes"
	"encoding/binary"
	"errors"
	"github.com/kuking/seof/crypto"
)

const HeaderMagic uint64 = 0xb0a713c
const HeaderLength int = 128

type Header struct {
	Magic         uint64
	ScriptSalt    [96]byte
	ScriptN       uint32
	ScriptR       uint32
	ScriptP       uint32
	DiskBlockSize uint32
	TailOfZeros   [8]byte
}

func (h *Header) Verify() error {
	if h.Magic != HeaderMagic {
		return errors.New("header: invalid magic")
	}
	if h.DiskBlockSize < 1112 || h.DiskBlockSize > 196608 {
		return errors.New("header: invalid disk_block_size")
	}
	for i := 0; i < len(h.ScriptSalt); i++ {
		if h.ScriptSalt[i] != 0 {
			break
		}
		if i == len(h.ScriptSalt)-1 {
			return errors.New("header: zero salt")
		}
	}

	if h.ScriptN > crypto.MaxSCryptParameters.N || h.ScriptN < crypto.MinSCryptParameters.N ||
		h.ScriptR > crypto.MaxSCryptParameters.R || h.ScriptR < crypto.MinSCryptParameters.R ||
		h.ScriptP > crypto.MaxSCryptParameters.P || h.ScriptP < crypto.MinSCryptParameters.P {
		return errors.New("header: invalid scrypt parameters")
	}

	for i := 0; i < len(h.TailOfZeros); i++ {
		if h.TailOfZeros[i] != 0 {
			return errors.New("header: tail of zeros not correct")
		}
	}

	return nil
}

type BlockEnvelop struct {
	Nonce         [nonceSize]byte
	CipherTextLen uint32
	CipherText    []byte
}

type BlockZero struct {
	BEncBlockSize uint32 //BEnc as 'Before Encryption'
	DiskBlockSize uint32
	BEncFileSize  uint64 //reported file size
	BlocksWritten uint64
}

func (z *BlockZero) Bytes() []byte {
	buf := new(bytes.Buffer)
	_ = binary.Write(buf, binary.LittleEndian, z)
	return buf.Bytes()
}

func BlockZeroFromBytes(b []byte) (*BlockZero, error) {
	r := bytes.NewReader(b)
	bz := BlockZero{}
	err := binary.Read(r, binary.LittleEndian, &bz)
	if err != nil {
		return nil, err
	}
	return &bz, nil
}


crypto/scrypt.go
package crypto

// Script parameters configuration valid at Dec 2020 for an AMD Ryzen 7 3800X 8-Core Processor
//
// Minimum accepted: a key spec with lower parameters shall not be accepted; processing in less than 21ms
// Currently recommended: parameters targeting at least 250ms for deriving a key
// Maximum accepted: an upper limit defined to avoid potential DoS attacks; maximum set at 4s

type SCryptParameters struct {
	N uint32
	R uint32
	P uint32
}

var RecommendedSCryptParameters = SCryptParameters{ //>600ms
	N: 1 << 16,
	R: 1 << 6,
	P: 1 << 0,
}

var BetterSCryptParameters = SCryptParameters{ //>5s
	N: 1 << 18,
	R: 1 << 7,
	P: 1 << 0,
}

var MaxSCryptParameters = SCryptParameters{ //>15s
	N: 1 << 18,
	R: 1 << 8,
	P: 1 << 0,
}

var MinSCryptParameters = SCryptParameters{ //21ms
	N: 1 << 14,
	R: 1 << 2,
	P: 1 << 0,
}


cli/birthday/main.go
package main

import (
	"fmt"
	"math"
	"math/big"
)

// The following calculates the probability of duplicating a nonce when picked at random (as the current implementation
// does). The current implementation uses 3 nonces of 12 bytes each (for each AES256/GCM cipher).
//
// The probability of picking the same nonce twice (birthday paradox) is calculated for one nonce of 12 bytes (96 bits).
//
// 100M, 0.999999999999936891126951232322755154461737176572294203867910474082896986505554273464571044238582672950519...
// lower than 0.00000000000001, or 1 in 100.000.000.000.000
// 100M blocks of 1K is approx. 95GiB written.
//
// 1258M, 0.99999999999001261703522929514448885033012604685025065706759067812784429829344821159506777018445782306879...
// lower than 0.000000000001, or  1 in 1.000.000.000.000
// 1258M blocks of 1K is approx. 1.17TiB written.
//
// 39805M, 0.9999999900007903323961553855263092697380031246078278537831527269974242016478462970070809347116269575979...
// lower than 0.000000001 or 1 in 1.000.000.000 (1 billion)
// 39805M block of 1K is approx. 37.07TiB written.
//
// You will have to write 37TiB into one file to have a chance of 1 in a billion to have a duplicated nonce, you also
// have to consider it would compromise the first layer of AES256/GCM, there will be two more layers to break.
//
// see: `birthday.log.xz` for the dump of this program after running for about 50hs

func main() {
	const prec = 5000   // bits
	const bits = 12 * 8 // the first nonce for the first AES256
	const n = 1_000_000_000_000

	days := new(big.Float).SetPrec(prec).SetFloat64(math.Pow(2, bits))

	fmt.Printf("Precision: %v mantissa bits\n", prec)
	fmt.Printf("Days: %.0f\n", days)
	fmt.Printf("N: %v\n", n)

	p := new(big.Float).SetPrec(prec).SetInt64(1)

	work := new(big.Float).SetPrec(prec)
	iFlt := new(big.Float).SetPrec(prec)
	one := new(big.Float).SetPrec(prec).SetInt64(1)
	for i := 0; i < n; i++ {
		// p *= (days - i) / days
		work.Sub(days, iFlt)
		work.Quo(work, days)
		p.Mul(p, work)
		iFlt.Add(iFlt, one)
		if i%1_000_000 == 0 {
			fmt.Printf("%vM, %.200f\n", i/1000/1000, p)
		}
	}
}


cli/seof/seof.go
package main

import (
	"encoding/hex"
	"flag"
	"fmt"
	pwe "github.com/kuking/go-pwentropy"
	"github.com/kuking/seof"
	"github.com/kuking/seof/crypto"
	"io"
	"io/ioutil"
	"os"
)

var doEncrypt bool
var passwordFile string
var doHelp bool
var doInfo bool
var blockSize uint
var scryptParamsCli string

func doArgsParsing() bool {
	flag.BoolVar(&doEncrypt, "e", false, "encrypt (default: to decrypt)")
	flag.StringVar(&scryptParamsCli, "scrypt", "default", "Encrypting Scrypt parameters: min, default, better, max")
	flag.BoolVar(&doInfo, "i", false, "show seof encrypted file metadata")
	flag.StringVar(&passwordFile, "p", "", "password file")
	flag.UintVar(&blockSize, "s", 1024, "block size")
	flag.BoolVar(&doHelp, "h", false, "Show usage")
	flag.Parse()
	if doHelp || flag.NArg() != 1 {
		fmt.Printf("Usage of %v: seof file utility\n\n", os.Args[0])
		flag.PrintDefaults()
		fmt.Print(`
NOTES: 
  - Password must be provided in a file. Command line is not secure in a multi-user host.
  - When encrypting, contents have to be provided via stdin pipe, decrypted output will be via stdout.
  - Scrypt parameters target times in modern CPUs (2021): min>20ms, default>600ms, better>5s, max>9s

Examples: 
  $ cat file | seof -e -p @password_file file.seof
  $ seof -p @password_file file.seof > file
  $ seof -i -p @password_file file.seof 
`)
		return false
	}
	return true
}

func main() {

	if !doArgsParsing() {
		os.Exit(-1)
	}

	if passwordFile == "" {
		_, _ = os.Stderr.WriteString("password not provided.\n")
		os.Exit(-1)
	}

	if len(passwordFile) > 1 && passwordFile[0] == '@' {
		passwordFile = passwordFile[1:]
	}
	passwordBytes, err := ioutil.ReadFile(passwordFile)
	if err != nil {
		panic(err)
	}
	password := string(passwordBytes)

	entropy := pwe.FairEntropy(password)
	if entropy < 96 {
		_, _ = os.Stderr.WriteString(fmt.Sprintf("FATAL: Est. entropy for provided password is not enough: %2.2f (minimum: 96)\n\n", entropy))
		password = pwe.PwGen(pwe.FormatEasy, pwe.Strength256)
		entropy = pwe.FairEntropy(password)
		_, _ = os.Stderr.WriteString(fmt.Sprintf("We have created a password for you with %2.2f bits of entropy \n"+
			"+-------------------------------------------------------+\n"+
			"| %52v  |\n"+
			"+-------------------------------------------------------+\n", entropy, password))
		os.Exit(-1)
	}

	var scryptParams crypto.SCryptParameters
	if doEncrypt {
		switch scryptParamsCli {
		case "min":
			scryptParams = crypto.MinSCryptParameters
		case "default":
			scryptParams = crypto.RecommendedSCryptParameters
		case "better":
			scryptParams = crypto.BetterSCryptParameters
		case "max":
			scryptParams = crypto.MaxSCryptParameters
		default:
			fmt.Println("SCrypt parameter not recognised:", scryptParamsCli)
			os.Exit(-1)
		}

	}

	filename := os.Args[len(os.Args)-1]
	var ef *seof.File
	if doInfo || !doEncrypt {
		ef, err = seof.OpenExt(filename, password, 10)
	} else {
		ef, err = seof.CreateExt(filename, password, scryptParams, int(blockSize), 10)
	}
	assertNoError(err, "Failed to open file: "+filename+" -- %v")

	if doInfo {
		stats, err := ef.Stat()
		assertNoError(err, "FATAL: problems doing file stats %v")

		fmt.Printf("           File Name: %v\n", stats.Name())
		fmt.Printf("   Modification Time: %v\n", stats.ModTime())
		fmt.Printf("           File Mode: %v \n", stats.Mode())
		fmt.Printf("        Content Size: %v bytes\n", stats.Size())
		fmt.Printf("   File Size On Disk: %v bytes\n", stats.EncryptedSize())
		fmt.Printf(" Encryption Overhead: %2.2f%%\n", float32(stats.EncryptedSize())*100/float32(stats.Size())-100)
		fmt.Printf("  Content Block Size: %v bytes\n", stats.BEBlockSize())
		fmt.Printf("Encrypted Block Size: %v bytes\n", stats.DiskBlockSize())
		fmt.Printf(" Total Blocks Writen: %v (= unique nonces)\n", stats.BlocksWritten())
		var scryptLevel string
		salt, n, r, p := stats.SCryptParameters()
		if n == crypto.MinSCryptParameters.N && r == crypto.MinSCryptParameters.R && p == crypto.MinSCryptParameters.P {
			scryptLevel = "Minimal (>20ms)"
		} else if n == crypto.MaxSCryptParameters.N && r == crypto.MaxSCryptParameters.R && p == crypto.MaxSCryptParameters.P {
			scryptLevel = "Maximum (>9s)"
		} else if n == crypto.RecommendedSCryptParameters.N && r == crypto.RecommendedSCryptParameters.R && p == crypto.RecommendedSCryptParameters.P {
			scryptLevel = "Recommended (>600ms)"
		} else if n == crypto.BetterSCryptParameters.N && r == crypto.BetterSCryptParameters.R && p == crypto.BetterSCryptParameters.P {
			scryptLevel = "Better (>5s)"
		} else {
			scryptLevel = "Unknown"
		}
		fmt.Printf("       SCrypt Preset: %v\n", scryptLevel)
		fmt.Printf("   SCrypt Parameters: N=%v, R=%v, P=%v, keyLength=96, salt=\n", n, r, p)
		hexa := hex.EncodeToString(salt)
		fmt.Printf("%69v\n%69v\n%69v\n", hexa[:64], hexa[64:128], hexa[128:])

	} else if doEncrypt {
		_, err = io.Copy(ef, os.Stdin)
	} else {
		_, err = io.Copy(os.Stdout, ef)
	}
	assertNoError(err, "FATAL: io error: %v")

	err = ef.Close()
	assertNoError(err, "FATAL: could not close the seof file: %v")
}

func assertNoError(err error, pattern string) {
	if err != nil {
		_, _ = os.Stderr.WriteString(fmt.Sprintf(pattern+"\n", err))
		os.Exit(-1)
	}
}
